File: 450_delete_node_in_a_BST.java
i/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return root;

        if (root.val < key) {
            root.right = deleteNode(root.right, key);
        } else if (root.val > key) {
            root.left = deleteNode(root.left, key);
        } else {
            // found node here
            if (root.left == null) {    // case 1: node has only right child
                root = root.right;
            } else if (root.right == null) {    // case 2: node has only left child
                root = root.left;
            } else {    // case 3: node has 2 children
                // find the left most child at right children
                // assign this value to current node
                root.val = findInorderSuccessor(root.right);

                // then delete the right node in the right children nodes
                root.right = deleteNode(root.right, root.val);
            }
        }
        return root;
    }

    private int findInorderSuccessor(TreeNode root) {
        int value = root.val;
        while (root.left != null) {
            value = root.left.val;
            root = root.left;
        }
        return value;
    }
}

File: python/450_delete_node_in_a_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # Time Complexity: O(logN)
    # Space Complexity: O(1)
    def minValueNode(self, node: TreeNode) -> TreeNode:
        curr = node
        while curr.left:
            curr = curr.left
        return curr

    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if not root:
            return root
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            if not root.left:
                temp = root.right
                root = None
                return temp
            elif not root.right:
                temp = root.left
                root = None
                return temp

            temp = self.minValueNode(root.right)

            # copy inorder successor's value to current node
            root.val = temp.val

            # delete the inorder successor
            root.right = self.deleteNode(root.right, temp.val)
        return root
